sycl/include/CL/sycl/half_type.hpp

```diff
@@ -85,13 +87,135 @@ inline float cast_if_host_half(half_impl::half val) {
+#endif
+
+// Partial specialization of `std::hash<cl::sycl::half>`
+template <> struct hash<HALF_TYPE> {
+  size_t operator()(HALF_TYPE const &Key) const noexcept {
+    return hash<uint16_t>()(reinterpret_cast<const uint16_t &>(Key));
   }
 };
 
+// Partial specialization of `std::numeric<cl::sycl::half>`
+
+// All following values are calculated based on description of each
+// function/value on https://en.cppreference.com/w/cpp/types/numeric_limits.
+
+// The following values are from cl_platform.h
+#define HLF_MIN 6.103515625e-05F
+
+#define HLF_MAX 65504.0F
+
+#define HLF_MAX_10_EXP 4
+
+#define HLF_MAX_EXP 16
+
+#define HLF_MIN_10_EXP -4
+
+#define HLF_MIN_EXP -13
+
+#define HLF_MANT_DIG 11
+
+#define HLF_DIG 3
+
+#define HLF_DECIMAL_DIG 5
+
+#define HLF_EPSILON 9.765625e-04F
+
+template <> struct numeric_limits<HALF_TYPE> {
+  static constexpr const bool is_specialized = true;
+
+  static constexpr const bool is_signed = true;
+
+  static constexpr const bool is_integer = false;
+
+  static constexpr const bool is_exact = false;
+
+  static constexpr const bool has_infinity = true;
+
+  static constexpr const bool has_quiet_NaN = true;
+
+  static constexpr const bool has_signaling_NaN = true;
+
+  static constexpr const float_denorm_style has_denorm = denorm_present;
+
+  static constexpr const bool has_denorm_loss = false;
+
+  static constexpr const bool tinyness_before = false;
+
+  static constexpr const bool traps = false;
+
+  static constexpr const int max_exponent10 = HLF_MAX_10_EXP;
+
+  static constexpr const int max_exponent = HLF_MAX_EXP;
+
+  static constexpr const int min_exponent10 = HLF_MIN_10_EXP;
+
+  static constexpr const int min_exponent = HLF_MIN_EXP;
+
+  static constexpr const int radix = 2;
+
+  static constexpr const int max_digits10 = HLF_DECIMAL_DIG;
+
+  static constexpr const int digits = HLF_MANT_DIG;
+
+  static constexpr const bool is_bounded = true;
+
+  static constexpr const int digits10 = HLF_DIG;
+
+  static constexpr const bool is_modulo = false;
+
+  static constexpr const bool is_iec559 = true;
+
+  static constexpr const float_round_style round_style = round_to_nearest;
+
+  static CONSTEXPR_QUALIFIER HALF_TYPE min() noexcept { return HLF_MIN; }
+
+  static CONSTEXPR_QUALIFIER HALF_TYPE max() noexcept { return HLF_MAX; }
+
+  static CONSTEXPR_QUALIFIER HALF_TYPE lowest() noexcept { return -HLF_MAX; }
+
+  static CONSTEXPR_QUALIFIER HALF_TYPE epsilon() noexcept {
+    return HLF_EPSILON;
+  }
+
+  static CONSTEXPR_QUALIFIER HALF_TYPE round_error() noexcept { return 0.5F; }
+
+  static CONSTEXPR_QUALIFIER HALF_TYPE infinity() noexcept {
+    return __builtin_huge_valf();
+  }
+
+  static CONSTEXPR_QUALIFIER HALF_TYPE quiet_NaN() noexcept {
+    return __builtin_nan("");
+  }
+
+  static CONSTEXPR_QUALIFIER HALF_TYPE signaling_NaN() noexcept {
+    return __builtin_nan("");
+  }
+
+  static CONSTEXPR_QUALIFIER HALF_TYPE denorm_min() noexcept {
+    return 5.96046e-08;
```

Comment:
Is this the exact value or do we should have an hexadecimal floating point value here to have the exact representation?

Reply:
Or like one suggestion above, to have `uint16_t` representation here and then `reinterpret_cast`?

Reply:
Just say no! For the sake of non obfuscated code, I would prefer the floating point value... :-)